fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,470 @@
+    else:
+        epoch_participation = state.previous_epoch_participation
+
+    return set([
+        i for i in active_validator_indices
+        if has_flag(epoch_participation[i], TIMELY_TARGET_FLAG_INDEX)
+    ])
+```
+
+#### `get_ffg_support`
+
+```python
+def get_ffg_support(store: Store, checkpoint: Root) -> Gwei:
+    """
+    Returns the total weight supporting the checkpoint in the block's chain at block's epoch.
+    """
+    current_epoch = get_current_store_epoch(store)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [checkpoint.epoch, checkpoint.epoch + 1]
+
+    if checkpoint not in store.checkpoint_states:
+        return Gwei(0)
+
+    checkpoint_state = store.checkpoint_states[checkpoint]
+
+    leaf_roots = [
+        leaf for leaf in get_leaf_block_roots(store, checkpoint.root)
+        if get_checkpoint_block(store, leaf, checkpoint.epoch) == checkpoint.root]
+
+    active_checkpoint_indices = get_active_validator_indices(checkpoint_state, checkpoint.epoch)
+    participating_indices_from_blocks = set().union(*[
+        get_epoch_participating_indices(
+            store.block_states[root],
+            active_checkpoint_indices,
+            checkpoint.epoch == current_epoch
+        )
+        for root in leaf_roots
+    ])
+
+    participating_indices_from_lmds = set([
+        i
+        for i in store.latest_messages
+        if get_checkpoint_block(store, store.latest_messages[i].root, store.latest_messages[i].epoch) == checkpoint.root
+    ])
+
+    return get_total_balance(checkpoint_state, participating_indices_from_blocks.union(participating_indices_from_lmds))
+```
+
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(store: Store, block_root: Root, epoch: Epoch) -> bool:
+    """
+    Returns whether the `block_root`'s checkpoint will be justified by the end of this epoch.
+    """
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+    checkpoint = Checkpoint(
+        epoch=epoch,
+        root=get_checkpoint_block(store, block_root, epoch)
+    )
+
+    store_target_checkpoint_state(store, checkpoint)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    total_active_balance = int(get_total_active_balance(store.checkpoint_states[checkpoint]))
+
+    ffg_support_for_checkpoint = int(get_ffg_support(store, checkpoint))
+
+    if epoch == current_epoch:
+        remaining_ffg_weight = int(get_remaining_weight_in_current_epoch(store, checkpoint))
+    else:
+        remaining_ffg_weight = 0
+
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            ceil_div(total_active_balance * CONFIRMATION_BYZANTINE_THRESHOLD, 100),
+            CONFIRMATION_SLASHING_THRESHOLD,
+            ffg_support_for_checkpoint
+        )
```

Comment:
Is ``ceil_div(total_active_balance * CONFIRMATION_BYZANTINE_THRESHOLD, 100)`` always smaller than or equal to  ``CONFIRMATION_SLASHING_THRESHOLD``?

Reply:
In the paper, it is ``ceil_div((total_active_balance - remaining_ffg_weight) * CONFIRMATION_BYZANTINE_THRESHOLD, 100)``. I wonder whether it is a typo.

Reply:
> Is ceil_div(total_active_balance * CONFIRMATION_BYZANTINE_THRESHOLD, 100) always smaller than or equal to CONFIRMATION_SLASHING_THRESHOLD?

Which paper are you referring to?